flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/aggregate/window/processors/UnsliceWindowAggProcessor.java

```diff
@@ -0,0 +1,391 @@
+    public TypeSerializer<TimeWindow> createWindowSerializer() {
+        return unsliceAssigner
+                .getMergingWindowAssigner()
+                .getWindowSerializer(new ExecutionConfig());
+    }
+
+    @Override
+    protected WindowTimerService<TimeWindow> getWindowTimerService() {
+        return new UnslicingWindowTimerServiceImpl(ctx.getTimerService(), shiftTimeZone);
+    }
+
+    private class WindowContextImpl
+            implements InternalWindowProcessFunction.Context<RowData, TimeWindow> {
+
+        @Override
+        public long currentProcessingTime() {
+            return ctx.getTimerService().currentProcessingTime();
+        }
+
+        @Override
+        public long currentWatermark() {
+            return ctx.getTimerService().currentWatermark();
+        }
+
+        @Override
+        public ZoneId getShiftTimeZone() {
+            return shiftTimeZone;
+        }
+
+        @Override
+        public RowData getWindowAccumulators(TimeWindow window) throws Exception {
+            return windowState.value(window);
+        }
+
+        @Override
+        public void setWindowAccumulators(TimeWindow window, RowData acc) throws Exception {
+            windowState.update(window, acc);
+        }
+
+        @Override
+        public void clearWindowState(TimeWindow window) throws Exception {
+            windowState.clear(window);
+            aggregator.cleanup(window);
+        }
+
+        @Override
+        public void clearPreviousState(TimeWindow window) throws Exception {}
+
+        @Override
+        public void clearTrigger(TimeWindow window) throws Exception {
+            triggerContext.setWindow(window);
+            triggerContext.clear();
+        }
+
+        @Override
+        public void deleteCleanupTimer(TimeWindow window) throws Exception {
+            long cleanupTime = toEpochMillsForTimer(window.maxTimestamp(), shiftTimeZone);
+            if (cleanupTime == Long.MAX_VALUE) {
+                // no need to clean up because we didn't set one
+                return;
+            }
+            if (unsliceAssigner.isEventTime()) {
+                triggerContext.deleteEventTimeTimer(cleanupTime);
+            } else {
+                triggerContext.deleteProcessingTimeTimer(cleanupTime);
+            }
+        }
+
+        @Override
+        public void onMerge(TimeWindow newWindow, Collection<TimeWindow> mergedWindows)
+                throws Exception {
+            triggerContext.setWindow(newWindow);
+            triggerContext.setMergedWindows(mergedWindows);
+            triggerContext.onMerge();
+        }
+
+        @Override
+        public <S extends State> S getPartitionedState(StateDescriptor<S, ?> stateDescriptor)
+                throws Exception {
+            requireNonNull(stateDescriptor, "The state properties must not be null");
+            return ctx.getKeyedStateBackend()
+                    .getPartitionedState(
+                            VoidNamespace.INSTANCE,
```

Comment:
Why the namespace is void?

Reply:
If the namespace is `Void`, that means the state gotten should not care about the namespace(window). In this case, this code is used for get state for `mapping` in `MergingWindowSet`. It does not pay attention to what the specific window is.